tools/OutlineService/OutlineService/OutlineService.cs

```diff
@@ -262,113 +264,136 @@ public partial class OutlineService : ServiceBase {
-      argsFormat = "interface ipv6 add route {0} interface={1} metric=0";
-      foreach (string subnet in IPV6_SUBNETS) {
-        RunCommand(CMD_NETSH, string.Format(argsFormat, subnet, NetworkInterface.IPv6LoopbackInterfaceIndex));
+      // Save the proxy's IP so we can reset routing.
+      this.proxyIp = proxyIp;
+
+      try {
+        // Route IPv4 traffic through the router. Instead of deleting the default IPv4 gateway (0.0.0.0/0),
+        // we resort to creating two more specific routes (see IPV4_SUBNETS) that take precedence over the
+        // default gateway. This way, we need not worry about the default gateway being recreated with a lower
+        // metric upon device sleep. This 'hack' was inspired by OpenVPN;
+        // see https://github.com/OpenVPN/openvpn3/commit/d08cc059e7132a3d3aee3dcd946fce4c35b1ced3#diff-1d76f0fd7ec04c6d1398288214a879c5R358.
+        foreach (string subnet in IPV4_SUBNETS) {
+          RunCommand(CMD_NETSH, $"interface ipv4 add route {subnet} nexthop={routerIp} interface={TAP_DEVICE_NAME} metric=0");
+        }
+      } catch (Exception e) {
+        throw new Exception($"could not change default gateway: {e.Message}");
       }
 
-      // TODO: inspect stderr, stdout for known errors (i.e. routes already exists); reset routing on failulre
-      return 0;
+      try {
+        // Outline does not currently support IPv6, so we resort to disabling it while the VPN is active to
+        // prevent leakage. Removing the default IPv6 gateway is not enough since it gets re-created
+        // through router advertisements and DHCP (disabling these or IPv6 routing altogether requires a
+        // system reboot). Thus, we resort to creating three IPv6 routes (see IPV6_SUBNETS) to the loopback
+        // interface that are more specific than the default route, causing IPv6 traffic to get dropped.
+        foreach (string subnet in IPV6_SUBNETS) {
+          RunCommand(CMD_NETSH, $"interface ipv6 add route {subnet} interface={NetworkInterface.IPv6LoopbackInterfaceIndex));} metric=0");
+        }
+      } catch (Exception e) {
+        throw new Exception($"could not disable IPv6: {e.Message}");
+      }
     }
 
-    // Resets the routing table to route traffic through the default IPv4 and IPv6 gatways.
-    public int ResetRouting(string proxyIp, string proxyInterfaceName) {
-      if (proxyIp == null) {
-        eventLog.WriteEntry("Asked to reset routing but we did not configure it!",
+    // Resets the routing table:
+    //  - remove route to the proxy server (if we know its IP)
+    //  - remove our default IPv4 gateways
+    //  - re-enable IPv6
+    //
+    // Does *not* throw or exit early if any step fails: keeps going!
+    public void ResetRouting(string proxyIp, string proxyInterfaceName) {
+      // Proxy server.
+      if (proxyIp != null) {
+        DeleteProxyRoute(proxyIp, proxyInterfaceName);

```

Comment:
Did you want to catch the exception here?

Reply:
Good catch. Now caught.